### ISSUE TITLE
Accordions Accessibility

### ISSUE THREAD

**Issue by scottaohara** [18F/web-design-standards#1121]
Hi there,

I just wanted to point out a couple of issues I've found with the [accordions example](https://playbook.cio.gov/designstandards/accordions/)

1a. If JavaScript is blocked or unavailable on a device, any accordion content not set to be displayed by default is completely inaccessible. A fix for this would be to dynamically add the aria-hidden* attributes when JavaScript is enabled, but to not include them when JavaScript is not enabled.

1b. Additionally, when JavaScript is disabled, the buttons obviously don't work, and in a no-js solution, these buttons should either not exist, or should actually be rendered as Headings to the accordion panels they are lead-ins to.

2. The accordion panels (.usa-accordion-content) would benefit from having an aria-labelledby attribute. This would require the buttons to have their own unique IDs as well.

*As an aside, resolving the (aria-hidden issue)[https://github.com/18F/web-design-standards/issues/1120] would have implications on this component, and any other component that requires aria-hidden to be set to display none.  For these sorts of examples, aria-hidden would need to be defined as display: none on a per-component basis.  Or, as I mentioned in the aria-hidden issue, these components could instead be updated to use [hidden] set to display none, as that would be the HTML native equivalent of aria-hidden="true" anyway.

Please let me know if there is any further information I can provide to help with this issue.

Thank you

**Comment by jhfrench** [comment]
To take this a little further--consider removing the button element entirely, and instead progressively-enhance `.usa-accordion [aria-controls]` ... something like (quick and dirty in jQuery)

```
$('.usa-accordion')
    //hide all but the first panel
    .find('.usa-accordion-content').not(':first').attr('aria-hidden, true').end().end()
    .find('[aria-controls]').attr('aria-role', '').on('click', function(){
        //show or hide the child panel associated with the clicked panel header
        $('#'+$(this).attr('aria-controls')).toggle(); 
    });
```

This approach provides a usable experience even without JavaScript (but a better experience with it).

**Comment by jhfrench** [comment]
You might also consider incorporating `role="tablist"`, `role="tab"`, and `role="tabpanel"` into the example HTML.

**Comment by scottaohara** [comment]
Agreed with all the above. 

there are also additional keyboard controls that should be implemented here, but I didn't this single issue to balloon out of control.

**Comment by msecret** [comment]
Thanks for all your comments! We've known about issue 1a, I think we have it tracked somewhere (although I can't find it right now so might need to make a new issue).

We're currently looking into the other concerns, including #1120 .

**Comment by msecret** [comment]
Ok created an issue #1122, please feel free to submit PRs for that, although the solution may be dependent on what we decided to do with #1120

**Comment by jpyuda** [comment]
Broadly speaking, I like an approach similar to @jhfrench's here -- default to the accordion being open, if javascript is available, close it, and write in the buttons. (In the past I've mostly handled this in CSS using a no-js class but there are plenty of different approaches; the main thing is to have the content available to anybody who doesn't have js for whatever reason).

An additional improvement that would help both usability and accessibility would be explicitly, visually labeling the expand/collapse button with "show" and "hide" or similar. Those ➕ and ➖ icons are somewhat ambiguous and labels would help a lot.

We might also want to consider switching to icons that suggest movement (similar to [CFPB's approach](http://cfpb.github.io/design-manual/ui-toolkit/expandables.html))

**Comment by maya** [comment]
@scottaohara @jhfrench Just wanted to note the `hidden` attribute is supported by IE 11 and above and this project supports IE 9 and above, so doesn't look like we can use it. http://caniuse.com/#search=hidden

**Comment by jhfrench** [comment]
@maya  True story, we still can't use it to reliably toggle visibility. So I edited [my April 14 code snippet](https://github.com/18F/web-design-standards/issues/1121#issuecomment-210011724) to accomodate and created [a working fiddle for demonstration purposes](https://jsfiddle.net/jhfrench/uv001vxL/).

Because [`aria-hidden` support is improving](https://www.paciellogroup.com/blog/2013/11/html5-accessibility-chops-hidden-aria-hidden-support/), I believe there is value in "setting" it.

**Comment by jhfrench** [comment]
@maya oh, hold on, misread that. You are talking about HTML5's `hidden`, I was talking about `aria-hidden`. I just checked my code to confirm that I do not make use of the `hidden` attribute, so what I write above still "stands".

**Comment by maya** [comment]
@jhfrench the note was in response to @scottaohara's comment:  

>Or, as I mentioned in the aria-hidden issue, these components could instead be updated to use [hidden] set to display none, as that would be the HTML native equivalent of aria-hidden="true" anyway.

I think `aria-hidden` gives us a lot more in terms of accessibility, so may be the preferred method.

**Comment by scottaohara** [comment]
hi @maya 

I was aware of ```[hidden]``` not being fully supported by itself until IE11, as I had mentioned in my original comment that the ```display: none !important;``` rule would be better set to the hidden attribute, as the following:

```css
[hidden] { 
  display: none; 
}
```

makes the attribute behave as expected in all browsers, and thus could have been a native solution for the ```aria-hidden``` attribute, without needing to define a new ```.display-none``` class.

Regardless, I've been following along with the updates in both of the related threads, and it looks like a solution has been decided upon.  I just didn't want to leave it out there that I was suggesting a fix that I hadn't actually thought through :)

**Comment by joshbruce** [comment]
I had a much longer response related directly to JS and the accordions, but I'm thinking @shawnbot and #1224 has most of the concerns and requirements pretty well covered. (I'll store my longer text elsewhere for a moment in case it's needed.)

I'm curious about the possibility for stuttering on load based on how many accordions are on the page.

**Aside re `hidden`**

@scottaohara does bring up an interesting point regarding a root-level attribute selector effectively giving IE9 support for the `hidden` attribute.

 Maybe something to consider in general - not just for the accordions? As in a separate "enhancement" issue could be opened to discuss further?

**Comment by shawnbot** [comment]
I'm not too concerned about the performance on load until there are hundreds of accordions on a single page, in which case I'd suggest that the author consider restructuring their content. 😉

**Comment by maya** [comment]
Thanks all for the fruitful discussion + @shawnbot for implementing this!

Closed via #1224.
